broadinstitute/single_cell_portal_core
# Storing all facets/filters from Azul service (SCP-3808)

In order to allow users to discover external data effectively via Azul, we need to pre-populate our `SearchFacet` entries will all possible facet/filter values from Azul.  This will enable users to select values from the filter lists that will correspond with an entry in Azul.  In addition, this update introduces three notable changes to executing searches:

1. A new attribute called `merged_filters` will store filter values from Azul alongside all entries from SCP studies (sourced from BigQuery)
2. An instance method on `SearchFacet` objects called `filters_for_user(user)` that will take into account the `cross_dataset_search_backend` feature flag
3. Changes to query string delimiters for search facets:
  a. Facet entries are now delimited with `|` instead of `+`
  b. Filter values are now delimited with `;` instead of `,`

Changes 1 & 2 are needed to prevent sign-in users from seeing a large list of filter values that will return empty searches, as they likely will not have the `cross_dataset_search_backend` feature flag enabled.  Change 3 is needed as filter values from Azul contain our current delimiters (`+` and `,`) within actual values (e.g. `CD4+ T cell` or `CD8-positive, alpha-beta T cell` for `cell_type`).  Changing our delimiter characters proved simpler than adding in complicated logic to double URL-encode/decode values.  As of right now, there are no entries in Azul that contain either `|` or `;` for a filter value.  If at some point such entries are added, this logic will need to be revisited.

MANUAL TESTING
1. Pull this branch, run migrations via `rails db:migrate` and boot as normal
2. On the home page, open the `disease` facet, and confirm it looks as normal for your instance (something like below)
![Screen Shot 2021-11-02 at 11 47 27 AM](https://user-images.githubusercontent.com/729968/139882223-337be70d-6a8c-4f9d-b0dc-64c65c6e8bbd.png)
3. Log in with an admin account, and ensure the `cross_dataset_search_backend` feature flag is enabled for this account
4. Go back to the home page and open the `disease` facet, and confirm that it has many more entries now:
![Screen Shot 2021-11-02 at 11 47 52 AM](https://user-images.githubusercontent.com/729968/139883037-40ff8da1-3049-40aa-a560-525b4b90574c.png)
 5. Select `Alzheimer disease` from the list, execute the search, and confirm one of the returned studies is `Single-cell RNA Sequencing of human microglia from post mortem Alzheimers Disease CNS tissue.`
 6. (Optional): log out, and log in with an account that does not have `cross_dataset_search_backend` enabled, and confirm the `disease` filter list has reverted back to normal

This PR satisfies SCP-3808.
